### PR TITLE
Adds try/catch for all network requests

### LINKED ISSUE #466: App crashes when doing any network requests on airplane mode

### PR DESCRIPTION
**Pull request by florina-muntenescu**
Fixes #466

### CODE REVIEW

File: core/src/main/java/io/plaidapp/core/designernews/data/comments/CommentsRemoteDataSource.kt [hunk 1]
```diff
@@ -31,18 +31,23 @@ class CommentsRemoteDataSource(private val service: DesignerNewsService) {
...
+            val response = service.getComments(requestIds).await()
+            if (response.isSuccessful) {
+                val body = response.body()
+                if (body != null) {
+                    return Result.Success(body)
+                }
             }
+            return Result.Error(IOException("Error getting comments ${response.code()} ${response.message()}"))
+        } catch (e: Exception) {
+            return Result.Error(IOException("Error getting comments  ${e.message}"))
```
Comment (nickbutcher) [inline core/src/main/java/io/plaidapp/core/designernews/data/comments/CommentsRemoteDataSource.kt]: Let's pass `e` as the cause to the `IOException` constructor.

Same throughout.

File: core/src/main/java/io/plaidapp/core/designernews/data/comments/CommentsRemoteDataSource.kt [hunk 2]
```diff
@@ -31,18 +31,23 @@ class CommentsRemoteDataSource(private val service: DesignerNewsService) {
...
         }
-        return Result.Error(IOException("Error getting comments ${response.code()} ${response.message()}"))
     }
 
     companion object {
-        @Volatile private var INSTANCE: CommentsRemoteDataSource? = null
+        @Volatile
```
Comment (nickbutcher) [inline core/src/main/java/io/plaidapp/core/designernews/data/comments/CommentsRemoteDataSource.kt]: revert.

File: core/src/main/java/io/plaidapp/core/designernews/data/comments/CommentsRemoteDataSource.kt [hunk 3]
```diff
@@ -31,18 +31,23 @@ class CommentsRemoteDataSource(private val service: DesignerNewsService) {
      */
     suspend fun getComments(ids: List<Long>): Result<List<CommentResponse>> {
         val requestIds = ids.joinToString(",")
-        val response = service.getComments(requestIds).await()
-        if (response.isSuccessful) {
-            val body = response.body()
-            if (body != null) {
-                return Result.Success(body)
+        try {
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/designernews/data/comments/CommentsRemoteDataSource.kt]: Why not introduce a function that takes the original API call incl error as function param and perform the try/catch in there, wrapping a potential exception in a `Result.Error`?

Something along the lines of this:
```
fun safeApiCall(message: String, call: () -> Result) {
    return try {
        call.invoke()
    } catch (e: Exception) {
        Result.Error(IOException("Error $message", e)
    }
}
```

This way the error handling code would be centralized.
The call site could then look like this:

```
safeApiCall("getting comments", getComments())
```

File: core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.plaidapp.core.util
+
+import io.plaidapp.core.data.Result
+import java.io.IOException
+
+suspend fun <T : Any> safeApiCall(message: String, call: suspend () -> Result<T>): Result<T> {
```
Comment (keyboardsurfer) [inline core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt]: Document

File: core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+package io.plaidapp.core.util
+
+import io.plaidapp.core.data.Result
+import java.io.IOException
+
+/**
+ * Wrap a suspending API [call] in try/catch. In case an exception is thrown, a [Result.Error] is
+ * created based on the [message].
+ */
+suspend fun <T : Any> safeApiCall(message: String, call: suspend () -> Result<T>): Result<T> {
```
Comment (nickbutcher) [inline core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt]: Can this be marked `inline`?

File: core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    return try {
+        call.invoke()
```
Comment (nickbutcher) [inline core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt]: Just `call()` would be cleaner.

File: core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt (hunk 5 above)
Comment (nickbutcher) [reply core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt]: Can we rename `message` to `errorMessage` to make its purpose clearer?

File: core/src/main/java/io/plaidapp/core/designernews/data/comments/CommentsRemoteDataSource.kt [hunk 7]
```diff
@@ -29,7 +30,12 @@ class CommentsRemoteDataSource(private val service: DesignerNewsService) {
     /**
      * Get a list of comments based on ids from Designer News API.
      */
-    suspend fun getComments(ids: List<Long>): Result<List<CommentResponse>> {
+    suspend fun getComments(ids: List<Long>) = safeApiCall(
+        message = "Error getting comments",
+        call = { requestGetComments(ids) }
```
Comment (nickbutcher) [inline core/src/main/java/io/plaidapp/core/designernews/data/comments/CommentsRemoteDataSource.kt]: I think that we should either move this outside of the parens **or** make `call` the first param as that is the expected case.

Comment (nickbutcher) [reply core/src/main/java/io/plaidapp/core/designernews/data/comments/CommentsRemoteDataSource.kt]: Same throughout.

File: core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    } catch (e: Exception) {
```
Comment (nickbutcher) [inline core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt]: Perhaps comment why we have to catch `Exception` here.

File: core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt (hunk 5 above)
Comment (florina-muntenescu) [reply core/src/main/java/io/plaidapp/core/util/NetworkUtils.kt]: it can't be inline because we're calling a suspend function

**Pull request merged**